src/main/java/io/synadia/flink/source/NatsJetStreamSourceReader.java

```diff
@@ -0,0 +1,157 @@
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+public class NatsJetStreamSourceReader<OutputT> implements SourceReader<OutputT, NatsSubjectSplit> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(NatsJetStreamSourceReader.class);
+
+    private final String id;
+    private final ConnectionFactory connectionFactory;
+    private final DeserializationSchema<OutputT> payloadDeserializer;
+    private final SourceReaderContext readerContext;
+    private final List<NatsSubjectSplit> subbedSplits;
+    private final FutureCompletingBlockingQueue<Message> messages;
+    private Connection connection;
+    private Subscription subscription;
+    private NatsConsumeOptions config;
+    private JetStream js;
+    private String subject;
+    private final Boundedness mode;
+    public NatsJetStreamSourceReader(String sourceId,
+                                     ConnectionFactory connectionFactory,
+                                     NatsConsumeOptions natsConsumeOptions,
+                                     DeserializationSchema<OutputT> payloadDeserializer,
+                                     SourceReaderContext readerContext,
+                                     String subject,
+                                     Boundedness mode) {
+        id = sourceId + "-" + Utils.generatePrefixedId(sourceId);
+        this.connectionFactory = connectionFactory;
+        this.payloadDeserializer = payloadDeserializer;
+        this.readerContext = checkNotNull(readerContext);
+        subbedSplits = new ArrayList<>();
+        messages = new FutureCompletingBlockingQueue<>();
+        this.config= natsConsumeOptions;
+        this.subject = subject;
+        this.mode = mode;
+    }
+
+
+    @Override
+    public void start() {
+        LOG.debug("{} | start", id);
+        try {
+            connection = connectionFactory.connect();
+            js = connection.jetStream();
+            PullSubscribeOptions pullOptions = PullSubscribeOptions.builder()
+                    .stream(config.getStreamName()).bind(true).durable(config.getConsumerName())
+                    .build();
+            subscription = js.subscribe(subject, pullOptions);
+        }
+        catch (IOException e) {
+            throw new FlinkRuntimeException(e);
+        } catch (JetStreamApiException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public InputStatus pollNext(ReaderOutput<OutputT> output) throws Exception {
+        List<Message> messages =
+                ((JetStreamSubscription) subscription).fetch(config.getBatchSize(), Duration.ofSeconds(5));
+        for (int i = 0; i < messages.size(); i++) {
+            Message message = messages.get(i);
+            boolean ackMessageFlag = (i == messages.size() - 1);
+            processMessage(output, message, ackMessageFlag);
```

Comment:
If a message is acknowledged but checkpointing fails, we lose the message.
I think message ack should be inside `notifyCheckpointComplete` method instead. Either that or inside `snapshotState`.

Reply:
Sure, will resolve this.

Reply:
Hi @vu-hoang-phan-form3, message is acknowledged in pollNext as we read, if for any reason checkpointing is failed after acknowledging the job for the message is anyways done, there is a chance of getting same message again if for some reason connector gets rebooted during message delivery and before it could ack the message, consumer will start redelivering message from where it stopped. 
If you can help me understand how do we lose the message here if checkpointing is failed ?

Reply:
@vu-hoang-phan-form3 you mean the processed data for the message might get lost?

Reply:
@shiv4289 yes
@sourabhaggrawal After `processMessage` here, the message is acknowledge and goes down the pipeline. Then if the next checkpoint fails, that same message will be lost. 

There is indeed a chance of redelivering the same message twice if it's acknowledged in `notifyCheckpointComplete`. That method is best-effort anyways. But it's not a problem, deduplication can be implemented elsewhere. Plus, NATS is at-least-once delivery, so duplication can always happen.

Reply:
@vu-hoang-phan-form3 @scottf can you please take a look at it once again, we have implemented checkpointing in this [PR](https://github.com/somratdutta/flink-connector-nats/pull/20/files)